File: lib/bindings/langs/android/lib/src/main/kotlin/breez_sdk_liquid_notification/job/LnurlPayInfo.kt
package breez_sdk_liquid_notification.job

import android.content.Context
import breez_sdk_liquid.BindingLiquidSdk
import breez_sdk_liquid_notification.Constants.DEFAULT_LNURL_PAY_INFO_NOTIFICATION_TITLE
import breez_sdk_liquid_notification.Constants.DEFAULT_LNURL_PAY_METADATA_PLAIN_TEXT
import breez_sdk_liquid_notification.Constants.DEFAULT_LNURL_PAY_NOTIFICATION_FAILURE_TITLE
import breez_sdk_liquid_notification.Constants.LNURL_PAY_INFO_NOTIFICATION_TITLE
import breez_sdk_liquid_notification.Constants.LNURL_PAY_METADATA_PLAIN_TEXT
import breez_sdk_liquid_notification.Constants.LNURL_PAY_NOTIFICATION_FAILURE_TITLE
import breez_sdk_liquid_notification.Constants.NOTIFICATION_CHANNEL_LNURL_PAY
import breez_sdk_liquid_notification.NotificationHelper.Companion.notifyChannel
import breez_sdk_liquid_notification.ResourceHelper.Companion.getString
import breez_sdk_liquid_notification.SdkForegroundService
import breez_sdk_liquid_notification.ServiceLogger
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

@Serializable
data class LnurlInfoRequest(
    @SerialName("callback_url") val callbackURL: String,
    @SerialName("reply_url") val replyURL: String,
)

// Serialize the response according to to LUD-06 payRequest base specification:
// https://github.com/lnurl/luds/blob/luds/06.md
@Serializable
data class LnurlPayInfoResponse(
    val callback: String,
    val maxSendable: ULong,
    val minSendable: ULong,
    val metadata: String,
    val tag: String,
)

class LnurlPayInfoJob(
    private val context: Context,
    private val fgService: SdkForegroundService,
    private val payload: String,
    private val logger: ServiceLogger,
) : LnurlPayJob {
    companion object {
        private const val TAG = "LnurlPayInfoJob"
    }

    override fun start(liquidSDK: BindingLiquidSdk) {
        var request: LnurlInfoRequest? = null
        try {
            request = Json.decodeFromString(LnurlInfoRequest.serializer(), payload)
            // Get the lightning limits
            val limits = liquidSDK.fetchLightningLimits()
            // Max millisatoshi amount LN SERVICE is willing to receive
            val maxSendableMsat = limits.receive.maxSat * 1000UL
            // Min millisatoshi amount LN SERVICE is willing to receive, can not be less than 1 or more than `maxSendableMsat`
            val minSendableMsat = limits.receive.minSat * 1000UL
            if (minSendableMsat < 1UL || minSendableMsat > maxSendableMsat) {
                throw InvalidLnurlPayException("Minimum sendable amount is invalid")
            }
            // Format the response
            val plainTextMetadata = getString(
                context, LNURL_PAY_METADATA_PLAIN_TEXT, DEFAULT_LNURL_PAY_METADATA_PLAIN_TEXT
            )
            val response = LnurlPayInfoResponse(
                request.callbackURL,
                maxSendableMsat,
                minSendableMsat,
                "[[\"text/plain\",\"$plainTextMetadata\"]]",
                "payRequest",
            )
            val success = replyServer(Json.encodeToString(response), request.replyURL)
            notifyChannel(
                context,
                NOTIFICATION_CHANNEL_LNURL_PAY,
                getString(
                    context,
                    if (success) LNURL_PAY_INFO_NOTIFICATION_TITLE else LNURL_PAY_NOTIFICATION_FAILURE_TITLE,
                    if (success) DEFAULT_LNURL_PAY_INFO_NOTIFICATION_TITLE else DEFAULT_LNURL_PAY_NOTIFICATION_FAILURE_TITLE
                ),
            )
        } catch (e: Exception) {
            logger.log(TAG, "Failed to process lnurl: ${e.message}", "WARN")
            if (request != null) {
                fail(e.message, request.replyURL)
            }
            notifyChannel(
                context,
                NOTIFICATION_CHANNEL_LNURL_PAY,
                getString(
                    context,
                    LNURL_PAY_NOTIFICATION_FAILURE_TITLE,
                    DEFAULT_LNURL_PAY_NOTIFICATION_FAILURE_TITLE
                ),
            )
        }

        fgService.onFinished(this)
    }

    override fun onShutdown() {}
}

File: lib/bindings/langs/swift/Sources/BreezSDKLiquid/Task/LnurlPayInfo.swift
import UserNotifications
import Foundation

struct LnurlInfoRequest: Codable {
    let callback_url: String
    let reply_url: String
}

struct LnurlInfoResponse: Decodable, Encodable {
    let callback: String
    let maxSendable: UInt64
    let minSendable: UInt64
    let metadata: String
    let commentAllowed: Int
    let tag: String
    let allowsNostr: Bool
    let nostrPubkey: String?

    init(
        callback: String, 
        maxSendable: UInt64, 
        minSendable: UInt64, 
        metadata: String, 
        commentAllowed: Int, 
        tag: String, 
        allowsNostr: Bool,
        nostrPubkey: String?
    ) {
        self.callback = callback
        self.maxSendable = maxSendable
        self.minSendable = minSendable
        self.metadata = metadata
        self.commentAllowed = commentAllowed
        self.tag = tag
        self.allowsNostr = allowsNostr
        self.nostrPubkey = nostrPubkey
    }
}

class LnurlPayInfoTask : LnurlPayTask {
    fileprivate let TAG = "LnurlPayInfoTask"

    init(payload: String, logger: ServiceLogger, contentHandler: ((UNNotificationContent) -> Void)? = nil, bestAttemptContent: UNMutableNotificationContent? = nil) {
        let successNotificationTitle = ResourceHelper.shared.getString(key: Constants.LNURL_PAY_INFO_NOTIFICATION_TITLE, fallback: Constants.DEFAULT_LNURL_PAY_INFO_NOTIFICATION_TITLE)
        let failNotificationTitle = ResourceHelper.shared.getString(key: Constants.LNURL_PAY_NOTIFICATION_FAILURE_TITLE, fallback: Constants.DEFAULT_LNURL_PAY_NOTIFICATION_FAILURE_TITLE)
        super.init(payload: payload, logger: logger, contentHandler: contentHandler, bestAttemptContent: bestAttemptContent, successNotificationTitle: successNotificationTitle, failNotificationTitle: failNotificationTitle)
    }

    override func start(liquidSDK: BindingLiquidSdk, pluginConfigs: PluginConfigs) throws {
        var request: LnurlInfoRequest? = nil
        do {
            request = try JSONDecoder().decode(LnurlInfoRequest.self, from: payload.data(using: .utf8)!)
        } catch let e {
            logger.log(tag: TAG, line: "Failed to decode payload: \(e)", level: "ERROR")
            displayPushNotification(title: failNotificationTitle, logger: logger, threadIdentifier: Constants.NOTIFICATION_THREAD_REPLACEABLE)
            throw e
        }

        do {
            logger.log(tag: TAG, line: "Fetching lightning limits", level: "INFO")
            let limits = try liquidSDK.fetchLightningLimits()

            // Max millisatoshi amount LN SERVICE is willing to receive
            let maxSendableMsat = limits.receive.maxSat * UInt64(1000)
            // Min millisatoshi amount LN SERVICE is willing to receive, can not be less than 1 or more than `maxSendableMsat`
            let minSendableMsat = limits.receive.minSat * UInt64(1000)
            if minSendableMsat < UInt64(1) || minSendableMsat > maxSendableMsat {
                logger.log(tag: TAG, line: "Invalid limits - minSendableMsat: \(minSendableMsat), maxSendableMsat: \(maxSendableMsat)", level: "ERROR")
                throw InvalidLnurlPayError.minSendable
            }
            // Format the response
            let plainTextMetadata = ResourceHelper.shared.getString(key: Constants.LNURL_PAY_METADATA_PLAIN_TEXT, fallback: Constants.DEFAULT_LNURL_PAY_METADATA_PLAIN_TEXT)
            let metadata = "[[\"text/plain\",\"\(plainTextMetadata)\"]]"
            logger.log(tag: TAG, line: "Sending info response", level: "INFO")
            let nostrPubkey = try? PluginManager.nwc(liquidSDK: liquidSDK, pluginConfigs: pluginConfigs)?.getInfo().walletPubkey;
            replyServer(encodable: LnurlInfoResponse(callback: request!.callback_url,
                                                     maxSendable: maxSendableMsat,
                                                     minSendable: minSendableMsat,
                                                     metadata: metadata,
                                                     commentAllowed: Constants.LNURL_PAY_COMMENT_MAX_LENGTH,
                                                     tag: "payRequest",
                                                     allowsNostr: nostrPubkey != nil,
                                                     nostrPubkey: nostrPubkey),
                        replyURL: request!.reply_url,
                        maxAge: Constants.CACHE_CONTROL_MAX_AGE_DAY)
        } catch let e {
            logger.log(tag: TAG, line: "Failed to process lnurl info: \(e)", level: "ERROR")
            fail(withError: e.localizedDescription, replyURL: request!.reply_url)
        }
    }
}
